polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2339,29 +2339,469 @@ class BinOpInfo {
+    return Result;
+
+  // If the lvalue is non-volatile, return the computed value of the
+  // assignment.
+  if (!LHS.isVolatile)
+    return Result;
+
+  llvm_unreachable("Volatile detection not implemented.");
+  return {LHS.getValue(builder), false};
+}
+
 #define HANDLEBINOP(OP)                                                        \
   ValueCategory MLIRScanner::VisitBin##OP(BinaryOperator *E) {                 \
     LLVM_DEBUG({                                                               \
       llvm::dbgs() << "VisitBin" #OP ": ";                                     \
       E->dump();                                                               \
       llvm::dbgs() << "\n";                                                    \
     });                                                                        \
-    return EmitBin##OP(EmitBinOps(E));                                         \
+    QualType PromotionType = Glob.getTypes().getPromotionType(E->getType());   \
+    ValueCategory Result = EmitBin##OP(EmitBinOps(E, PromotionType));          \
+    if (Result.val && !PromotionType.isNull())                                 \
+      Result = EmitUnPromotedValue(Result, E->getType());                      \
+    return Result;                                                             \
+  }                                                                            \
+                                                                               \
+  ValueCategory MLIRScanner::VisitBin##OP##Assign(BinaryOperator *E) {         \
+    LLVM_DEBUG({                                                               \
+      llvm::dbgs() << "VisitBin" #OP "Assign: ";                               \
+      E->dump();                                                               \
+      llvm::dbgs() << "\n";                                                    \
+    });                                                                        \
+    return EmitCompoundAssign(cast<CompoundAssignOperator>(E),                 \
+                              &MLIRScanner::EmitBin##OP);                      \
   }
 #include "Expressions.def"
 #undef HANDLEBINOP
 
-BinOpInfo MLIRScanner::EmitBinOps(BinaryOperator *E) {
-  const ValueCategory LHS = Visit(E->getLHS());
-  const ValueCategory RHS = Visit(E->getRHS());
-  const QualType Ty = E->getType();
+BinOpInfo MLIRScanner::EmitBinOps(BinaryOperator *E, QualType PromotionType) {
+  const ValueCategory LHS = EmitPromotedScalarExpr(E->getLHS(), PromotionType);
+  const ValueCategory RHS = EmitPromotedScalarExpr(E->getRHS(), PromotionType);
+  const QualType Ty = !PromotionType.isNull() ? PromotionType : E->getType();
   const BinaryOperator::Opcode Opcode = E->getOpcode();
   return {LHS, RHS, Ty, Opcode, E};
 }
 
 ValueCategory MLIRScanner::EmitBinMul(const BinOpInfo &Info) {
-  assert(Info.getOpcode() == BinaryOperator::Opcode::BO_Mul &&
-         "Invalid binary expression");
```

Comment:
Drop these assertions as Opcode might be coming from compound assignment operator.